---
#    self_contained: yes
output:
  github_document:
  html_preview: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.width = 13,
  fig.height = 9
)

set.seed(42)

unlink("dating_LD_files", recursive = TRUE)
unlink(list.files(".", "dating_LD.*.pdf"), recursive = TRUE)
```

```{r}
library(data.table)
library(ggplot2)
library(readr)
library(dplyr)
library(tidyr)
library(cowplot)

library(plyranges)
library(ggbio)
library(BSgenome.Hsapiens.UCSC.hg19)

library(slendr)
init_env()

suppressPackageStartupMessages(source(here::here("utils.R")))
```

Parameters from the simulation:

```{r}
gen_time <- 27
t_admix <- round(55000 / gen_time) * gen_time
```

## Read simulated admixture tracts

```{r, eval=file.exists("data/sim_tracts.tsv")}
tracts_df <-
  read_tsv("data/sim_tracts.tsv") %>%
  group_by(name) %>%
  mutate(haplotype = dense_rank(node_id),
         chrom = paste0("chr", haplotype)) %>%
  ungroup()

tracts_df
```

```{r}
tracts_gr <- tracts_df %>%
  makeGRangesFromDataFrame(
    start.field = "left",
    end.field = "right",
    keep.extra.columns = TRUE,
    ignore.strand = TRUE,
    starts.in.df.are.0based = TRUE
  )
```

```{r}
seqlengths(tracts_gr) <- rep(100e6, 2)
```

```{r}
tracts_gr
```

## Admixture dating using regularly-spaced loci

### Define regularly spaced ancestry-informative sites

```{r}
interval <- 10e3

sites_grl <- generate_info_sites(tracts_gr, interval = interval)
distances <- seq(interval, 1e6, by = interval)

file <- "dating_LD_pairs_regular.rds"
if (file.exists(file)) {
  pairs <- readRDS(file)
} else {
  tstart <- Sys.time()

  pairs <- collect_pairs(sites_grl[1], distances)
  pairs <- lapply(seqlevels(sites_grl), function(chr) pairs[[1]])
  names(pairs) <- seqlevels(sites_grl)
  
  tend <- Sys.time()

  print(tend - tstart)
  saveRDS(pairs, file)
}
```


### Compute covariance in all individuals

```{r}
length_cutoff <- 0

cov_reg_df <- compute_tract_covariances(filter(tracts_gr, length > length_cutoff), sites_grl, pairs)
```

```{r}
cov_reg_df %>%
ggplot() +
  geom_line(aes(distance, covariance, color = factor(sample_age), group = interaction(chrom, name))) +
  facet_grid(~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances)), ylim = c(0, 0.1)) +
  theme(legend.position = "bottom")
```

```{r}
tracts_gr %>%
filter(length > length_cutoff) %>%
as_tibble %>%
ggplot() +
  geom_density(aes(length, color = factor(sample_age), group = name)) +
  facet_grid(. ~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances)))
```

```{r}
chrom <- "chr1"
name <- "EUR_10"

sites_gr <- sites_grl[seqlevels(sites_grl) == chrom, ] %>% unlist
ind_tracts_gr <- tracts_gr %>% filter(name == !!name, seqnames == chrom)
ind_sites_gr <- sites_gr

# mark sites falling within an introgressed tract
tract_overlaps <- queryHits(findOverlaps(ind_sites_gr, ind_tracts_gr))
mcols(ind_sites_gr)$neand <- FALSE
mcols(ind_sites_gr[tract_overlaps])$neand <- TRUE
mcols(ind_sites_gr)$neand <- as.integer(mcols(ind_sites_gr)$neand)
    
ggplot() +
  geom_rect(data = as.data.frame(ind_tracts_gr), aes(xmin = start, xmax = end, ymin = 0.1, ymax = 1.5), fill = "blue") +
  geom_point(data = as.data.frame(filter(ind_sites_gr, neand == 1)), aes(x = start, y = 0.05), size = 0.1, color = "black") +
  geom_point(data = as.data.frame(ind_sites_gr), aes(x = start, y = 0), size = 0.1, alpha = 0.05) +
  scale_x_continuous(labels = scales::comma) +
  expand_limits(x = 0) +
  theme_minimal() +
  theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(), panel.grid = element_blank())
```



### Fitting exponential distribution to covariances

<!-- ```{r} -->
<!-- sample_age <- 40000 -->
<!-- name <- sample(filter(cov_df, sample_age == !!sample_age)$name, 1) -->
<!-- chrom <- sample(filter(cov_df, sample_age == !!sample_age)$chrom, 1) -->

<!-- gen_time <- 27 -->
<!-- r <- 1e-8 -->

<!-- data_df <- filter(cov_df, name == !!name, chrom == !!chrom) #sample_age == !!sample_age) -->

<!-- # nls_res <- nls(covariance ~ exp(-lambda * distance * r) + c,  -->
<!-- #                data = data_df,  start = list(A = 1, lambda = 1/50000, c = 0)) -->

<!-- nls_res <- nls(covariance ~ SSasymp(distance, Asym, R0, lrc), data = data_df) -->

<!-- summary(nls_res) -->

<!-- plot(data_df$distance, data_df$covariance) -->
<!-- y_nls <- predict(nls_res, newdata = data_df[, c("sample_age", "distance")]) %>% as.vector -->
<!-- lines(data_df$distance, y_nls) -->

<!-- lambda_nls <- exp(coef(nls_res)["lrc"]) -->
<!-- (t_gens_nls <- lambda_nls / r) -->
<!-- (t_nls <- t_gens_nls * gen_time + sample_age) -->
<!-- ``` -->

```{r}
fit_reg_df <- fit_exponential(cov_reg_df)
```

```{r}
sample_age <- 30000
name <- sample(filter(cov_reg_df, sample_age == !!sample_age)$name, 1)

ind_cov_df <- filter(cov_reg_df, name == !!name)
ind_fit_df <- filter(fit_reg_df, name == !!name)

ggplot() +
  geom_line(data = ind_cov_df, aes(distance, covariance)) +
  geom_line(data = ind_fit_df, aes(distance, covariance), color = "red", linetype = "dashed") +
  facet_wrap(~ chrom)
```

```{r}
fit_reg_df %>%
ggplot(aes(factor(sample_age), t_admix, group = sample_age)) +
  geom_boxplot() +
  geom_point() +
  geom_hline(yintercept = t_admix, linetype = "dashed") +
  coord_cartesian(ylim = c(0, 100e3)) +
  theme_minimal() +
  theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank())
```



## Admixture dating using genotype data (tracts)

### Define archaic ancestry-informative sites

```{r}
model <- read_model("data/dating_model")
ts <- ts_load("data/dating.trees", model = model)

gt <- ts_genotypes(ts) %>% setDT()
names(gt) <- gsub("_chr", "_hap", names(gt))
```

```{r}
fixed_afr <- gt[, rowMeans(.SD) == 0, .SDcols = patterns("^AFR")]
fixed_neand <- gt[, rowMeans(.SD) == 1, .SDcols = patterns("^NEA")]
fixed_sites <- fixed_afr & fixed_neand

gt <- gt[fixed_sites, .SD, .SDcols = !patterns("^AFR")]
sites_ir <- IRanges(start = gt$pos, end = gt$pos, index = 1:sum(fixed_sites))
```

```{r}
sites_grl <- GRangesList(list(GRanges(seqnames = "chr1", sites_ir), GRanges(seqnames = "chr2", sites_ir)))
seqlevels(sites_grl) <- seqlevels(tracts_gr)
seqlengths(sites_grl) <- seqlengths(tracts_gr)
```

```{r}
interval <- 10e3
distances <- seq(interval, 1e6, by = interval)

file <- "dating_LD_pairs_mutations.rds"
if (file.exists(file)) {
  pairs <- readRDS(file)
} else {
  tstart <- Sys.time()

  pairs <- collect_pairs(sites_grl[1], distances)
  pairs <- lapply(seqlevels(sites_grl), function(chr) pairs[[1]])
  names(pairs) <- seqlevels(sites_grl)
  
  tend <- Sys.time()

  print(tend - tstart)
  saveRDS(pairs, file)
}
```

```{r}
length_cutoff <- 0

cov_mut_df <- compute_tract_covariances(filter(tracts_gr, length > length_cutoff), sites_grl, pairs)
```

```{r}
cov_mut_df %>%
ggplot() +
  geom_line(aes(distance, covariance, color = factor(sample_age), group = interaction(chrom, name))) +
  facet_grid(~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances)), ylim = c(0, 0.1)) +
  theme(legend.position = "bottom")
```

### Fitting exponential distribution to covariances

```{r}
fit_mut_df <- fit_exponential(cov_mut_df)
```

```{r}
sample_age <- 0
name <- sample(filter(fit_mut_df, sample_age == !!sample_age)$name, 1)

ind_cov_df <- filter(cov_mut_df, name == !!name)
ind_fit_df <- filter(fit_mut_df, name == !!name)

ggplot() +
  geom_line(data = ind_cov_df, aes(distance, covariance)) +
  geom_line(data = ind_fit_df, aes(distance, covariance), color = "red", linetype = "dashed") +
  facet_wrap(~ chrom)
```

```{r}
fit_mut_df %>%
ggplot(aes(factor(sample_age), t_admix, group = sample_age)) +
  geom_boxplot() +
  geom_point() +
  geom_hline(yintercept = t_admix, linetype = "dashed") +
  coord_cartesian(ylim = c(0, 100e3)) +
  theme_minimal() +
  theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank())
```

## Dating admixture using genotype data (info sites)

### Define archaic ancestry-informative sites

```{r}
model <- read_model("data/dating_model")
ts <- ts_load("data/dating.trees", model = model)
metadata <- ts_samples(ts) %>% dplyr::rename(sample_age = time)

gt <- ts_genotypes(ts) %>% setDT()
names(gt) <- gsub("_chr", "_hap", names(gt))
```

```{r}
fixed_afr <- gt[, rowMeans(.SD) == 0, .SDcols = patterns("^AFR")]
fixed_neand <- gt[, rowMeans(.SD) == 1, .SDcols = patterns("^NEA")]
fixed_sites <- fixed_afr & fixed_neand

gt <- gt[fixed_sites, .SD, .SDcols = !patterns("^AFR")]
sites_ir <- IRanges(start = gt$pos, end = gt$pos, index = 1:sum(fixed_sites))
```

```{r}
sites_grl <- GRangesList(list(GRanges(seqnames = "chr1", sites_ir), GRanges(seqnames = "chr2", sites_ir)))
seqlevels(sites_grl) <- seqlevels(tracts_gr)
seqlengths(sites_grl) <- seqlengths(tracts_gr)
```

```{r}
interval <- 10e3
distances <- seq(interval, 1e6, by = interval)

file <- "dating_LD_pairs_mutations.rds"
if (file.exists(file)) {
  pairs <- readRDS(file)
} else {
  tstart <- Sys.time()

  pairs <- collect_pairs(sites_grl[1], distances)
  pairs <- lapply(seqlevels(sites_grl), function(chr) pairs[[1]])
  names(pairs) <- seqlevels(sites_grl)
  
  tend <- Sys.time()

  print(tend - tstart)
  saveRDS(pairs, file)
}
```

```{r}
# split individual haplotypes into separate "chromosomes"
# (i.e. an individual's haplotypes "X_1_chr1" and "X_1_chr2" along "chr1",
# which are produced by ts_genotypes() will be transformed into X_1 column
# along two chromosomes "chr1", and "chr2" -- this confusing terminology
# will be later fixed in slendr's ts_genotypes() so that it produces 
# columns such as "X_1_hap1" and "X_1_hap2" instead)
info_gt <- lapply(seqlevels(sites_grl), function(chrom) {
  hap <- gsub("chr", "", chrom)

  new_gt <- copy(gt)
  new_gt[, chrom := chrom]

  ind_names <- grep(paste0("_hap", hap, "$"), names(gt), value = TRUE)
  cols <- c("chrom", "pos", ind_names)
  new_gt <- new_gt[, .SD, .SDcols = cols]

  names(new_gt) <- gsub("_hap\\d", "", names(new_gt))

  new_gt
}) %>%
  do.call(rbind, .)

cov_info_df <- compute_match_covariances(info_gt, pairs, metadata)
```

```{r}
cov_info_df %>%
ggplot() +
  geom_line(aes(distance, covariance, color = factor(sample_age), group = interaction(chrom, name))) +
  facet_grid(~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances)), ylim = c(0, 0.1)) +
  theme(legend.position = "bottom")
```

### Fitting exponential distribution to covariances

```{r}
fit_info_df <- fit_exponential(cov_info_df)
```

```{r}
sample_age <- 50000
name <- sample(filter(fit_mut_df, sample_age == !!sample_age)$name, 1)

ind_cov_df <- filter(cov_info_df, name == !!name)
ind_fit_df <- filter(fit_info_df, name == !!name)

ggplot() +
  geom_line(data = ind_cov_df, aes(distance, covariance)) +
  geom_line(data = ind_fit_df, aes(distance, covariance), color = "red", linetype = "dashed") +
  facet_wrap(~ chrom)
```

```{r}
fit_info_df %>%
ggplot(aes(factor(sample_age), t_admix, group = sample_age)) +
  geom_boxplot() +
  geom_point() +
  geom_hline(yintercept = t_admix, linetype = "dashed") +
  coord_cartesian(ylim = c(0, 100e3)) +
  theme_minimal() +
  theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank())
```

## Comparison of exponential fits

```{r}
cov_df <- rbind(
  mutate(cov_reg_df, method = "regular sites (using tracts)"),
  mutate(cov_mut_df, method = "natural sites (using tracts)"),
  mutate(cov_info_df, method = "natural sites (without tracts)")
) %>% mutate(method = factor(method, levels = c("regular sites (using tracts)",
                                                "natural sites (using tracts)",
                                                "natural sites (without tracts)")))
```

```{r}
cov_df %>%
group_by(method, sample_age, distance) %>%
summarise(covariance = mean(covariance)) %>%
ggplot() +
  geom_line(aes(distance, covariance, color = factor(method))) +
  facet_grid(~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances))) +
  theme(legend.position = "bottom")
```


```{r}
fit_df <- rbind(
  mutate(fit_reg_df, method = "regular sites (using tracts)"),
  mutate(fit_mut_df, method = "natural sites (using tracts)"),
  mutate(fit_info_df, method = "natural sites (without tracts)")
) %>% mutate(method = factor(method, levels = c("regular sites (using tracts)",
                                                "natural sites (using tracts)",
                                                "natural sites (without tracts)")))
```

```{r}
fit_df %>%
ggplot(aes(factor(sample_age), t_admix, group = sample_age)) +
  geom_boxplot(aes(fill = factor(sample_age))) +
  geom_point() +
  geom_hline(yintercept = t_admix, linetype = "dashed") +
  coord_cartesian(ylim = c(0, 100e3)) +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank()) +
  facet_wrap(~ method)
```

## Empirical data (using only informative sites)


### Read inferred ancestry tracts

```{r}
metadata <- read_metadata() %>% mutate(sample_age = ageAverage, name = sampleId)
tracts_df <- rbind(read_tracts("Modern"), read_tracts("Ancient"))

tracts_df <- select(metadata, name = sampleId, sample_age = ageAverage, coverage) %>%
  inner_join(tracts_df, by = c("name" = "ID")) %>%
  dplyr::rename(left = start, right = end) %>%
  filter(chrom == "chr21")
```

```{r}
tracts_gr <- tracts_df %>%
  makeGRangesFromDataFrame(
    start.field = "left",
    end.field = "right",
    keep.extra.columns = TRUE,
    ignore.strand = TRUE,
    starts.in.df.are.0based = TRUE
  )
```

```{r}
seqlengths(tracts_gr) <- seqlengths(BSgenome.Hsapiens.UCSC.hg19)[names(seqlengths(tracts_gr))]
genome(tracts_gr) <- "hg19"
```

### Read genotypes at informative sites

```{r}
info_gt <- fread("info_gt_chr21.tsv.gz")

gt_samples <- unique(gsub("_hap\\d$", "", colnames(info_gt)))
meta_samples <- metadata$sampleId
both_samples <- intersect(meta_samples, gt_samples)
info_gt <- info_gt[, .SD, .SDcols = c("chrom", "pos", paste0(both_samples, "_hap1"), paste0(both_samples, "_hap2"), "NEA_1")]

sites_gr <- makeGRangesFromDataFrame(info_gt[, .(chrom, pos)], start.field = "pos", end.field = "pos")
mcols(sites_gr)$index <- 1:nrow(info_gt)
```

```{r}
sites_grl <- GRangesList(list(sites_gr))
seqlevels(sites_grl) <- seqlevels(tracts_gr)
seqlengths(sites_grl) <- seqlengths(tracts_gr)
```

```{r}
interval <- 10e3
distances <- seq(interval, 1e6, by = interval)

file <- "dating_LD_pairs_empirical.rds"
if (file.exists(file)) {
  pairs <- readRDS(file)
} else {
  tstart <- Sys.time()

  pairs <- collect_pairs(sites_grl, distances)
  names(pairs) <- seqlevels(sites_grl)
  
  tend <- Sys.time()

  print(tend - tstart)
  saveRDS(pairs, file)
}
```

```{r}
cov_emp_df <- compute_match_covariances(info_gt, pairs, metadata) %>%
  mutate(hap = gsub(".*_hap(\\d)", "\\1", name))
```

<!-- ```{r} -->
<!-- cov_emp_df %>% -->
<!-- ggplot() + -->
<!--   geom_line(aes(distance, covariance, color = factor(sample_age), group = interaction(chrom, name))) + -->
<!--   facet_grid(~ sample_age) + -->
<!--   coord_cartesian(xlim = c(0, max(distances)), ylim = c(0, 0.1)) + -->
<!--   theme(legend.position = "bottom") -->
<!-- ``` -->

### Fitting exponential distribution to covariances

```{r}
fit_emp_df <- fit_exponential(cov_emp_df) %>%
  mutate(hap = gsub(".*_hap(\\d)", "\\1", name))
```

```{r}
sample_age <- sample(fit_emp_df$sample_age, 1)
name <- sample(filter(fit_emp_df, sample_age == !!sample_age)$name, 1) %>%
  gsub("_hap\\d$", "", .)

ind_cov_df <- filter(cov_emp_df, grepl(!!name, name)) %>% mutate(name = gsub("_hap\\d", "", name))
ind_fit_df <- filter(fit_emp_df, grepl(!!name, name)) %>% mutate(name = gsub("_hap\\d", "", name))

ggplot() +
  geom_line(data = ind_cov_df, aes(distance, covariance)) +
  geom_line(data = ind_fit_df, aes(distance, covariance), color = "red", linetype = "dashed") +
  facet_wrap(hap ~ chrom)
```

```{r}
fit_emp_df %>%
ggplot(aes(sample_age, t_admix, color = sample_age)) +
  geom_point() +
  geom_hline(yintercept = t_admix, linetype = "dashed") +
  coord_cartesian(ylim = c(0, 100e3)) +
  theme_minimal() +
  theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank())
```
