---
#    self_contained: yes
output:
  github_document:
  html_preview: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.width = 13,
  fig.height = 9
)

set.seed(42)

unlink("dating_LD_files", recursive = TRUE)
unlink(list.files(".", "dating_LD.*.pdf"), recursive = TRUE)
```

```{r}
library(data.table)
library(ggplot2)
library(readr)
library(dplyr)
library(tidyr)
library(cowplot)

library(plyranges)
library(ggbio)
library(BSgenome.Hsapiens.UCSC.hg19)

library(slendr)
init_env()

suppressPackageStartupMessages(source(here::here("utils.R")))
```

Parameters from the simulation:

```{r}
gen_time <- 27
t_admix <- round(55000 / gen_time) * gen_time
```

## Read simulated admixture tracts

```{r, eval=file.exists("data/sim_tracts.tsv")}
tracts_df <-
  read_tsv("data/sim_tracts.tsv") %>%
  group_by(name) %>%
  mutate(haplotype = dense_rank(node_id),
         chrom = paste0("chr", haplotype)) %>%
  ungroup()

tracts_df
```

<!-- ```{r} -->
<!-- metadata <- read_metadata() -->
<!-- tracts_df <- rbind(read_tracts("Modern"), read_tracts("Ancient")) -->

<!-- tracts_df <- select(metadata, name = sampleId, ageAverage, coverage) %>% -->
<!--   inner_join(tracts_df, by = c("name" = "ID")) %>% -->
<!--   dplyr::rename(left = start, right = end) -->

<!-- tracts_df <- tracts_df %>% filter(name == "Kostenki") %>% arrange(chrom, left, right) -->
<!-- tracts_df$chrom_id <- tracts_df$chrom -->
<!-- ``` -->

```{r}
tracts_gr <- tracts_df %>%
  makeGRangesFromDataFrame(
    start.field = "left",
    end.field = "right",
    keep.extra.columns = TRUE,
    ignore.strand = TRUE,
    starts.in.df.are.0based = TRUE
  )
```

<!-- ```{r} -->
<!-- seqlengths(tracts_gr) <- seqlengths(BSgenome.Hsapiens.UCSC.hg19)[names(seqlengths(tracts_gr))] -->
<!-- genome(tracts_gr) <- "hg19" -->
<!-- ``` -->

```{r}
seqlengths(tracts_gr) <- rep(100e6, 2)
```

```{r}
tracts_gr
```

## Admixture dating using regularly-spaced loci

### Define regularly spaced ancestry-informative sites

```{r}
interval <- 10e3

sites_grl <- generate_info_sites(tracts_gr, interval = interval)
distances <- seq(interval, 1e6, by = interval)

file <- "dating_LD_pairs_tracts_regular.rds"
if (file.exists(file)) {
  pairs <- readRDS(file)
} else {
  tstart <- Sys.time()

  pairs <- collect_pairs(sites_grl[1], distances)
  pairs <- lapply(seqlevels(sites_grl), function(chr) pairs[[1]])
  names(pairs) <- seqlevels(sites_grl)
  
  tend <- Sys.time()

  print(tend - tstart)
  saveRDS(pairs, file)
}
```


### Compute covariance in a given individual

```{r}
length_cutoff <- 50e3

cov_reg_df <- compute_tract_covariances(filter(tracts_gr, length > length_cutoff), sites_grl, pairs)
```

<!-- ```{r} -->
<!-- cov_df <- cov_ind_df %>%  group_by(sample_age, distance) %>% summarise(covariance = mean(covariance)) -->
<!-- ``` -->

```{r}
cov_reg_df %>%
ggplot() +
  geom_line(aes(distance, covariance, color = factor(sample_age), group = interaction(chrom, name))) +
  facet_grid(~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances))) +
  theme(legend.position = "bottom")
```

```{r}
tracts_gr %>%
filter(length > length_cutoff) %>%
as_tibble %>%
ggplot() +
  geom_density(aes(length, color = factor(sample_age), group = name)) +
  facet_grid(. ~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances)))
```

```{r}
chrom <- "chr1"
name <- "EUR_10"

sites_gr <- sites_grl[seqlevels(sites_grl) == chrom, ] %>% unlist
ind_tracts_gr <- tracts_gr %>% filter(name == !!name, seqnames == chrom)
ind_sites_gr <- sites_gr

# mark sites falling within an introgressed tract
tract_overlaps <- queryHits(findOverlaps(ind_sites_gr, ind_tracts_gr))
mcols(ind_sites_gr)$neand <- FALSE
mcols(ind_sites_gr[tract_overlaps])$neand <- TRUE
mcols(ind_sites_gr)$neand <- as.integer(mcols(ind_sites_gr)$neand)
    
ggplot() +
  geom_rect(data = as.data.frame(ind_tracts_gr), aes(xmin = start, xmax = end, ymin = 0.1, ymax = 1.5), fill = "blue") +
  geom_point(data = as.data.frame(filter(ind_sites_gr, neand == 1)), aes(x = start, y = 0.05), size = 0.1, color = "black") +
  geom_point(data = as.data.frame(ind_sites_gr), aes(x = start, y = 0), size = 0.1, alpha = 0.05) +
  scale_x_continuous(labels = scales::comma) +
  expand_limits(x = 0) +
  theme_minimal() +
  theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(), panel.grid = element_blank())
```



### Fitting exponential distribution to covariances

<!-- ```{r} -->
<!-- sample_age <- 40000 -->
<!-- name <- sample(filter(cov_df, sample_age == !!sample_age)$name, 1) -->
<!-- chrom <- sample(filter(cov_df, sample_age == !!sample_age)$chrom, 1) -->

<!-- gen_time <- 27 -->
<!-- r <- 1e-8 -->

<!-- data_df <- filter(cov_df, name == !!name, chrom == !!chrom) #sample_age == !!sample_age) -->

<!-- # nls_res <- nls(covariance ~ exp(-lambda * distance * r) + c,  -->
<!-- #                data = data_df,  start = list(A = 1, lambda = 1/50000, c = 0)) -->

<!-- nls_res <- nls(covariance ~ SSasymp(distance, Asym, R0, lrc), data = data_df) -->

<!-- summary(nls_res) -->

<!-- plot(data_df$distance, data_df$covariance) -->
<!-- y_nls <- predict(nls_res, newdata = data_df[, c("sample_age", "distance")]) %>% as.vector -->
<!-- lines(data_df$distance, y_nls) -->

<!-- lambda_nls <- exp(coef(nls_res)["lrc"]) -->
<!-- (t_gens_nls <- lambda_nls / r) -->
<!-- (t_nls <- t_gens_nls * gen_time + sample_age) -->
<!-- ``` -->

```{r}
grid_df <- expand_grid(chrom = seqlevels(tracts_gr), name = unique(tracts_gr$name))

fit_df <- lapply(1:nrow(grid_df), function(i) {
  name <- grid_df[i, ]$name
  chrom <- grid_df[i, ]$chrom

  data_df <- filter(cov_reg_df, name == !!name, chrom == !!chrom)
  nls_res <- nls(covariance ~ SSasymp(distance, Asym, R0, lrc), data = data_df)

  df <- tibble(
    distance = data_df$distance,
    covariance = predict(nls_res, newdata = data_df[, "distance"])
  )

  r <- 1e-8

  tibble(
    name = name,
    chrom = chrom,
    sample_age = data_df$sample_age[1],
    lambda = exp(coef(nls_res)["lrc"]),
    t_gens_before = lambda / r,
    t_admix = t_gens_before * gen_time + sample_age,
    fit = list(df)
  )
}) %>%
  do.call(rbind, .) %>%
  unnest(fit)
```

```{r}
sample_age <- 30000
name <- sample(filter(cov_reg_df, sample_age == !!sample_age)$name, 1)

ind_cov_df <- filter(cov_reg_df, name == !!name)
ind_fit_df <- filter(fit_df, name == !!name)

ggplot() +
  geom_line(data = ind_cov_df, aes(distance, covariance)) +
  geom_line(data = ind_fit_df, aes(distance, covariance), color = "red", linetype = "dashed") +
  facet_wrap(~ chrom)
```

```{r}
fit_df %>%
ggplot(aes(factor(sample_age), t_admix, group = sample_age)) +
  geom_boxplot() +
  geom_point() +
  geom_hline(yintercept = t_admix, linetype = "dashed") +
  coord_cartesian(ylim = c(0, 100e3)) +
  theme_minimal() +
  theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank())
```


<!-- ## Dating admixture using genotype data -->

<!-- ```{r} -->
<!-- model <- read_model("dating_model") -->
<!-- ts <- ts_load("dating.trees", model = model) -->

<!-- gt <- ts_genotypes(ts) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- afr_gt <- select(gt, starts_with("AFR")) -->
<!-- nea_gt <- select(gt, starts_with("NEA")) -->

<!-- fixed_afr <- rowMeans(afr_gt) == 0 -->
<!-- fixed_nea <- rowMeans(nea_gt) == 1 -->

<!-- fixed_sites <- fixed_afr & fixed_nea -->

<!-- info_gt <- gt[fixed_sites, ] -->
<!-- info_ir <- IRanges(start = info_gt$pos, end = info_gt$pos, index = 1:sum(fixed_sites)) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- sites_grl <- GRangesList(list(GRanges(seqnames = "chr1_1", info_ir), GRanges(seqnames = "chr1_2", info_ir))) -->
<!-- seqlevels(sites_grl) <- seqlevels(tracts_gr) -->
<!-- seqlengths(sites_grl) <- seqlengths(tracts_gr) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- interval <- 10e3 -->
<!-- distances <- seq(interval, 1e6, by = interval) -->

<!-- # unlink("dating_LD_pairs_mutations.rds") -->
<!-- if (file.exists("dating_LD_pairs_mutations.rds")) { -->
<!--   pairs <- readRDS("dating_LD_pairs_mutations.rds") -->
<!-- } else { -->
<!--   tstart <- Sys.time() -->

<!--   pairs <- collect_pairs(sites_grl[1], distances) -->
<!--   pairs <- lapply(seqlevels(sites_grl), function(chr) pairs[[1]]) -->
<!--   names(pairs) <- seqlevels(sites_grl) -->

<!--   tend <- Sys.time() -->

<!--   print(tend - tstart) -->
<!--   saveRDS(pairs, "dating_LD_pairs_mutations.rds") -->
<!-- } -->
<!-- ``` -->

<!-- ```{r} -->
<!-- length_cutoff <- 50e3 -->

<!-- cov_mut_df <- compute_tract_covariances(filter(tracts_gr, length > length_cutoff), sites_grl, pairs) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- cov_mut_df %>% -->
<!-- ggplot() + -->
<!--   geom_line(aes(distance, covariance, color = factor(sample_age), group = interaction(chrom, name))) + -->
<!--   facet_grid(~ sample_age) + -->
<!--   coord_cartesian(xlim = c(0, max(distances))) + -->
<!--   theme(legend.position = "bottom") -->
<!-- ``` -->

<!-- ```{r} -->
<!-- tracts_gr %>% -->
<!-- filter(length > length_cutoff) %>% -->
<!-- as_tibble %>% -->
<!-- ggplot() + -->
<!--   geom_density(aes(length, color = factor(sample_age), group = name)) + -->
<!--   facet_grid(. ~ sample_age) + -->
<!--   coord_cartesian(xlim = c(0, max(distances))) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- chrom <- "chr1_2" -->
<!-- name <- "EUR_10" -->

<!-- sites_gr <- sites_grl[seqlevels(sites_grl) == chrom, ] %>% unlist -->
<!-- ind_tracts_gr <- tracts_gr %>% filter(name == !!name, seqnames == chrom) -->
<!-- ind_sites_gr <- sites_gr -->

<!-- # mark sites falling within an introgressed tract -->
<!-- tract_overlaps <- queryHits(findOverlaps(ind_sites_gr, ind_tracts_gr)) -->
<!-- mcols(ind_sites_gr)$neand <- FALSE -->
<!-- mcols(ind_sites_gr[tract_overlaps])$neand <- TRUE -->
<!-- mcols(ind_sites_gr)$neand <- as.integer(mcols(ind_sites_gr)$neand) -->

<!-- ggplot() + -->
<!--   geom_rect(data = as.data.frame(ind_tracts_gr), aes(xmin = start, xmax = end, ymin = 0.1, ymax = 1.5), fill = "blue") + -->
<!--   geom_point(data = as.data.frame(filter(ind_sites_gr, neand == 1)), aes(x = start, y = 0.05), size = 0.1, color = "black") + -->
<!--   geom_point(data = as.data.frame(ind_sites_gr), aes(x = start, y = 0), size = 0.1, alpha = 0.05) + -->
<!--   scale_x_continuous(labels = scales::comma) + -->
<!--   expand_limits(x = 0) + -->
<!--   theme_minimal() + -->
<!--   theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(), panel.grid = element_blank()) -->
<!-- ``` -->




## Dating admixture using genotype data (tracts)

```{r}
model <- read_model("dating_model")
ts <- ts_load("dating.trees", model = model)

gt <- ts_genotypes(ts) %>% setDT()
names(gt) <- gsub("_chr", "_hap", names(gt))
```

```{r}
fixed_afr <- gt[, rowMeans(.SD) == 0, .SDcols = patterns("^AFR")]
fixed_neand <- gt[, rowMeans(.SD) == 1, .SDcols = patterns("^NEA")]
fixed_sites <- fixed_afr & fixed_neand

pos <- gt[fixed_sites, pos]
info_gt <- gt[fixed_sites]
info_gt[, pos := NULL]
sites_ir <- IRanges(start = pos, end = pos, index = 1:sum(fixed_sites))
```

```{r}
sites_grl <- GRangesList(list(GRanges(seqnames = "chr1", sites_ir), GRanges(seqnames = "chr2", sites_ir)))
seqlevels(sites_grl) <- seqlevels(tracts_gr)
seqlengths(sites_grl) <- seqlengths(tracts_gr)
```

```{r}
interval <- 10e3
distances <- seq(interval, 1e6, by = interval)

file <- "dating_LD_pairs_tracts_mutations.rds"
if (file.exists(file)) {
  pairs <- readRDS(file)
} else {
  tstart <- Sys.time()

  pairs <- collect_pairs(sites_grl[1], distances)
  pairs <- lapply(seqlevels(sites_grl), function(chr) pairs[[1]])
  names(pairs) <- seqlevels(sites_grl)
  
  tend <- Sys.time()

  print(tend - tstart)
  saveRDS(pairs, file)
}
```

```{r}
length_cutoff <- 50e3

cov_mut_df <- compute_tract_covariances(filter(tracts_gr, length > length_cutoff), sites_grl, pairs)
```

```{r}
cov_mut_df %>%
ggplot() +
  geom_line(aes(distance, covariance, color = factor(sample_age), group = interaction(chrom, name))) +
  facet_grid(~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances))) +
  theme(legend.position = "bottom")
```

```{r}
tracts_gr %>%
filter(length > length_cutoff) %>%
as_tibble %>%
ggplot() +
  geom_density(aes(length, color = factor(sample_age), group = name)) +
  facet_grid(. ~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances)))
```

```{r}
chrom <- "chr1"
name <- "EUR_10"

sites_gr <- sites_grl[seqlevels(sites_grl) == chrom, ] %>% unlist
ind_tracts_gr <- tracts_gr %>% filter(name == !!name, seqnames == chrom)
ind_sites_gr <- sites_gr

# mark sites falling within an introgressed tract
tract_overlaps <- queryHits(findOverlaps(ind_sites_gr, ind_tracts_gr))
mcols(ind_sites_gr)$neand <- FALSE
mcols(ind_sites_gr[tract_overlaps])$neand <- TRUE
mcols(ind_sites_gr)$neand <- as.integer(mcols(ind_sites_gr)$neand)
    
ggplot() +
  geom_rect(data = as.data.frame(ind_tracts_gr), aes(xmin = start, xmax = end, ymin = 0.1, ymax = 1.5), fill = "blue") +
  geom_point(data = as.data.frame(filter(ind_sites_gr, neand == 1)), aes(x = start, y = 0.05), size = 0.1, color = "black") +
  geom_point(data = as.data.frame(ind_sites_gr), aes(x = start, y = 0), size = 0.1, alpha = 0.05) +
  scale_x_continuous(labels = scales::comma) +
  expand_limits(x = 0) +
  theme_minimal() +
  theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(), panel.grid = element_blank())
```



## Dating admixture using genotype data (info sites)

```{r}
model <- read_model("dating_model")
ts <- ts_load("dating.trees", model = model)
metadata <- ts_samples(ts) %>% dplyr::rename(sample_age = time)

gt <- ts_genotypes(ts) %>% setDT()
names(gt) <- gsub("_chr", "_hap", names(gt))
```

```{r}
fixed_afr <- gt[, rowMeans(.SD) == 0, .SDcols = patterns("^AFR")]
fixed_neand <- gt[, rowMeans(.SD) == 1, .SDcols = patterns("^NEA")]
fixed_sites <- fixed_afr & fixed_neand

gt <- gt[fixed_sites]
sites_ir <- IRanges(start = gt$pos, end = gt$pos, index = 1:sum(fixed_sites))
```

```{r}
sites_grl <- GRangesList(list(GRanges(seqnames = "chr1", sites_ir), GRanges(seqnames = "chr2", sites_ir)))
seqlevels(sites_grl) <- seqlevels(tracts_gr)
seqlengths(sites_grl) <- seqlengths(tracts_gr)
```

```{r}
interval <- 10e3
distances <- seq(interval, 1e6, by = interval)

file <- "dating_LD_pairs_info_mutations.rds"
if (file.exists(file)) {
  pairs <- readRDS(file)
} else {
  tstart <- Sys.time()

  pairs <- collect_pairs(sites_grl[1], distances)
  pairs <- lapply(seqlevels(sites_grl), function(chr) pairs[[1]])
  names(pairs) <- seqlevels(sites_grl)
  
  tend <- Sys.time()

  print(tend - tstart)
  saveRDS(pairs, file)
}
```

```{r}
length_cutoff <- 50e3

info_gt <- lapply(seqlevels(sites_grl), function(chrom) {
  new_gt <- copy(gt)
  new_gt[, chrom := chrom]
  new_gt
}) %>%
  do.call(rbind, .)

cov_info_df <- compute_match_covariances(info_gt, pairs, metadata)
```

```{r}
cov_info_df %>%
ggplot() +
  geom_line(aes(distance, covariance, color = factor(sample_age), group = interaction(chrom, name))) +
  facet_grid(~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances))) +
  theme(legend.position = "bottom")
```

```{r}
tracts_gr %>%
filter(length > length_cutoff) %>%
as_tibble %>%
ggplot() +
  geom_density(aes(length, color = factor(sample_age), group = name)) +
  facet_grid(. ~ sample_age) +
  coord_cartesian(xlim = c(0, max(distances)))
```

```{r}
chrom <- "chr1"
name <- "EUR_10"

sites_gr <- sites_grl[seqlevels(sites_grl) == chrom, ] %>% unlist
ind_tracts_gr <- tracts_gr %>% filter(name == !!name, seqnames == chrom)
ind_sites_gr <- sites_gr

# mark sites falling within an introgressed tract
tract_overlaps <- queryHits(findOverlaps(ind_sites_gr, ind_tracts_gr))
mcols(ind_sites_gr)$neand <- FALSE
mcols(ind_sites_gr[tract_overlaps])$neand <- TRUE
mcols(ind_sites_gr)$neand <- as.integer(mcols(ind_sites_gr)$neand)
    
ggplot() +
  geom_rect(data = as.data.frame(ind_tracts_gr), aes(xmin = start, xmax = end, ymin = 0.1, ymax = 1.5), fill = "blue") +
  geom_point(data = as.data.frame(filter(ind_sites_gr, neand == 1)), aes(x = start, y = 0.05), size = 0.1, color = "black") +
  geom_point(data = as.data.frame(ind_sites_gr), aes(x = start, y = 0), size = 0.1, alpha = 0.05) +
  scale_x_continuous(labels = scales::comma) +
  expand_limits(x = 0) +
  theme_minimal() +
  theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(), panel.grid = element_blank())
```
